src/workerd/io/container-client.c++

```diff
@@ -0,0 +1,277 @@
+
+  kj::Promise<void> connect(ConnectContext context) override {
+    kj::HttpHeaderTable headerTable;
+    kj::HttpHeaders headers(headerTable);
+    auto request = network.connect(
+        kj::str(containerHost, ":", containerPort), headers, kj::HttpConnectSettings());
+    auto upPipe = kj::newOneWayPipe();
+    auto upEnd = kj::mv(upPipe.in);
+    auto results = context.getResults();
+    results.setUp(byteStreamFactory.kjToCapnp(kj::mv(upPipe.out)));
+    auto downEnd = byteStreamFactory.capnpToKj(context.getParams().getDown());
+    co_await kj::joinPromisesFailFast(
+        kj::arr(upEnd->pumpTo(*request.connection), request.connection->pumpTo(*downEnd)));
+  }
+
+ private:
+  capnp::ByteStreamFactory& byteStreamFactory;
+  kj::HttpClient& network;
+  [[maybe_unused]] kj::String containerId;
+  kj::String containerHost;
+  uint16_t containerPort;
+};
+
+kj::Promise<ContainerClient::Response> ContainerClient::dockerApiRequest(
+    kj::HttpMethod method, kj::StringPtr endpoint, kj::Maybe<kj::StringPtr> body) {
+  kj::HttpHeaderTable::Builder builder;
+  const auto userAgent = builder.add("User-Agent");
+  const auto accept = builder.add("Accept");
+  const auto contentType = builder.add("Content-Type");
+  const auto contentLength = builder.add("Content-Length");
+  auto headerTable = builder.build();
+  kj::HttpHeaders headers(*headerTable);
+  headers.set(kj::HttpHeaderId::HOST, "localhost");
+  headers.set(userAgent, "workerd/1.0");
+  headers.set(accept, "*/*");
+
+  KJ_IF_SOME(requestBody, body) {
+    headers.set(contentType, "application/json");
+    headers.set(contentLength, kj::str(requestBody.size()));
+
+    auto req = httpClient->request(method, endpoint, headers, requestBody.size());
+    {
+      auto body = kj::mv(req.body);
+      co_await body->write(requestBody.asBytes());
+    }
+    auto response = co_await req.response;
+    auto result = co_await response.body->readAllText();
+    co_return Response{.statusCode = response.statusCode, .body = kj::mv(result)};
+  } else {
+    auto req = httpClient->request(method, endpoint, headers);
+    { auto body = kj::mv(req.body); }
+    auto response = co_await req.response;
+    auto result = co_await response.body->readAllText();
+    co_return Response{.statusCode = response.statusCode, .body = kj::mv(result)};
+  }
+}
+
+kj::Promise<bool> ContainerClient::isContainerRunning() {
+  // Docker API v1.50: GET /containers/{id}/json
+  auto endpoint = kj::str("/containers/", containerName, "/json");
+
+  auto response = co_await dockerApiRequest(kj::HttpMethod::GET, endpoint);
+  if (response.statusCode == 404) {
+    co_return false;
+  }
+  JSG_REQUIRE(response.statusCode == 200, Error, "Container inspect failed");
+
+  // Parse JSON response
+  capnp::JsonCodec codec;
+  codec.handleByAnnotation<docker_api::Docker::ContainerInspectResponse>();
+  capnp::MallocMessageBuilder message;
+  auto jsonRoot = message.initRoot<docker_api::Docker::ContainerInspectResponse>();
```

Comment:
Cf my earlier comment.

May dockerApiRequest should be templated and decode the JSON